modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/IgniteStatisticsManagerImpl.java

```diff
@@ -87,195 +142,420 @@ public IgniteStatisticsRepository statisticsRepository() {
 
     /** {@inheritDoc} */
     @Override public ObjectStatistics getLocalStatistics(String schemaName, String objName) {
-        return statsRepos.getLocalStatistics(new StatsKey(schemaName, objName));
+        return statsRepos.getLocalStatistics(new StatisticsKey(schemaName, objName));
+    }
+
+    /**
+     * Clear object statistics implementation.
+     *
+     * @param keys Keys to clear statistics by.
+     * @throws IgniteCheckedException In case of errors.
+     */
+    private void clearObjectStatistics(Collection<StatisticsKeyMessage> keys) throws IgniteCheckedException {
+        statCrawler.sendClearStatisticsAsync(keys);
+        /*Collection<StatisticsAddrRequest<StatisticsClearRequest>> clearReqs = helper.generateClearRequests(keys);
+
+        Collection<StatisticsAddrRequest<StatisticsClearRequest>> failedReqs = sendRequests(clearReqs);
+        if (!F.isEmpty(failedReqs))
+            if (log.isInfoEnabled())
+                log.info(String.format("Unable to send all statistics clear requests to %d nodes for keys %s",
+                    failedReqs.size(), keys));
+
+
+        UUID locId = ctx.localNodeId();
+        StatisticsAddrRequest<StatisticsClearRequest> locMsg = clearReqs.stream().filter(m -> locId.equals(m.targetNodeId())).findAny()
+                .orElse(null);
+        if (null != locMsg)
+            for (StatisticsKeyMessage locKey : locMsg.req().keys())
+                clearObjectStatisticsLocal(locKey);*/
+    }
+
+    /**
+     * Update counter to mark that statistics by some partitions where collected by remote request.
+     *
+     * @param gatId Gathering id.
+     * @param parts Partitions count.
+     */
+    public void onRemoteGatheringSend(UUID gatId, int parts) {
+        currColls.compute(gatId, (k,v) -> {
+           if (v == null) {
+               if (log.isDebugEnabled())
+                   log.debug(String.format("Unable to mark %d partitions gathered by gathering id %s", parts, gatId));
+
+               return null;
+           }
+
+           return v.registerCollected(Collections.emptyMap(), parts) ? null : v;
+        });
+    }
+
+    /** {@inheritDoc} */
+    @Override public void clearObjectStatistics(
+            final GridTuple3<String, String, String[]>... keys
```

Comment:
let's use class with meaningful name instead of GridTuple3

Reply:
I create StatisticsTarget for it.